fs.go

```diff
@@ -196,121 +221,163 @@ func (tfs *tarfs) Sub(dir string) (fs.FS, error) {
 	return subfs, nil
 }
 
-func (tfs *tarfs) get(name, op string) (fs.DirEntry, error) {
-	if !fs.ValidPath(name) {
-		return nil, &fs.PathError{Op: op, Path: name, Err: fs.ErrInvalid}
+func (tfs *tarfs) get(op, path string) (entry, error) {
+	if !fs.ValidPath(path) {
+		return nil, newErr(op, path, fs.ErrInvalid)
 	}
 
-	e, ok := tfs.entries[name]
+	e, ok := tfs.entries[path]
 	if !ok {
-		return nil, &fs.PathError{Op: op, Path: name, Err: fs.ErrNotExist}
+		return nil, newErrNotExist(op, path)
 	}
 
-	return e, nil
+	return e.(entry), nil
 }
 
-type entries interface {
-	append(fs.DirEntry)
-	get() []fs.DirEntry
+type entry interface {
+	fs.DirEntry
+	size() int64
+	readdir(path string) ([]fs.DirEntry, error)
+	readfile(path string) ([]byte, error)
+	entries(op, path string) ([]fs.DirEntry, error)
+	open() (*file, error)
 }
 
-type fileInfo interface {
-	FileInfo() (fs.FileInfo, error)
+type regEntry struct {
+	fs.DirEntry
+	name   string
+	ra     io.ReaderAt
+	offset int64
 }
 
-type entry struct {
-	h       *tar.Header
-	b       []byte
-	entries []fs.DirEntry
-}
-
-var _ fs.DirEntry = &entry{}
+var _ entry = &regEntry{}
 
-func (e *entry) Name() string {
-	return e.h.FileInfo().Name()
+func (e *regEntry) size() int64 {
+	info, _ := e.Info() // err is necessarily nil
+	return info.Size()
 }
 
-func (e *entry) IsDir() bool {
-	return e.h.FileInfo().IsDir()
+func (e *regEntry) readdir(path string) ([]fs.DirEntry, error) {
+	return nil, newErrNotDir("readdir", path)
 }
 
-func (e *entry) Type() fs.FileMode {
-	return e.h.FileInfo().Mode() & fs.ModeType
+func (e *regEntry) readfile(path string) ([]byte, error) {
+	tr := tar.NewReader(io.NewSectionReader(e.ra, e.offset, 1<<63-1))
+
+	if _, err := tr.Next(); err != nil {
+		return nil, err
+	}
+
+	b := bytes.NewBuffer(make([]byte, 0, e.size()))
+
+	if _, err := io.Copy(b, tr); err != nil {
+		return nil, err
+	}
+
+	return b.Bytes(), nil
 }
 
-func (e *entry) Info() (fs.FileInfo, error) {
-	return e.h.FileInfo(), nil
+func (e *regEntry) entries(op, path string) ([]fs.DirEntry, error) {
+	return nil, newErrNotDir(op, path)
 }
 
-var _ entries = &entry{}
+func (e *regEntry) open() (*file, error) {
+	b, err := e.readfile("")
+	if err != nil {
+		return nil, err
+	}
 
-func (e *entry) append(c fs.DirEntry) {
-	e.entries = append(e.entries, c)
+	return &file{e, bytes.NewReader(b), -1, false}, nil
```

Comment:
Is it possible to avoid copying the entire file into memory?

Are things structured this way solely because of the need to support sparse tars? How does that actually work? Does `tar.Reader.Read()` know how to stop reading separately from using the size reported by the tar file header?

Is it just a matter of keeping track of the begin and end offsets between headers, and then building an io.MultiReader for that `tar.Reader` data section, plus another reader that just reads 0's up to the rest of the difference between the length of that section and the size reported by the tar header?

Reply:
Scratch my last suggestion.

Looking more into the `tar.Reader` implementation, it looks like the returned `header.Size` is the actual number of bytes that can be read (including all of the extra 0 bytes that are not actually present in the tar file).

So can `open` do the following:
1. Create a new `tar.Reader` just for this entry
2. Call `tr.Next()` to prep the upcoming read
3. Return `tr`

Then `readfile` could call `open` and then return `io.ReadAll(tr)`

Reply:
Sorry for my stream of consciousness here...

Looks like the copy is necessary to appease `http.FS`? If the copies are otherwise unnecessary, what do you think about making the `io.Seeker` functionality optional?

Reply:
I rebased #16 and added a commit that implements optionally disabling the seeker implementation. Under the hood, disabling the seeker implementation results in no extra file copies when opening the file.

Reply:
As you guessed, the `io.Seeker` implementation is "just" for `http.FS`, however it could be useful in many other cases.

I agree that reading the whole file just to implement `io.Seeker` is a bad solution. 
I feel like your solution is the best right now.
The fake `io.Seeker` feels like a fake good idea, it will probably work for `http.FS`, but will fail in any other cases where an actual `io.Seeker` is needed.

Regarding your comment about keeping in v1 or bumping to v2 and changing the default behavior,  I'm not decided yet...

I'd like to try some more things this evening, and then we can make a decision. 

PS: Sorry for the delay, I'm on vacation.

Reply:
So I wanted to try reimplementing `io.Seeker` over the `tar.Reader`, with this approach:
 - if the Seek goes forward, read and discard in order to reach the new offset
 - if the Seek goes backward, recreate the `tar.Reader`, and if necessary read and discard in order to reach the new offset

You can see the changes here 30ec627f956427ff22442a561f72d7a22181d032

Tests are OK (`io.Seeker` implementation testing is covered by the `testing/fstest` of stdlib).
Benchmarks didn't suffer visible changes.

I'll integrate the new benchmarks you wrote tomorrow in order to measure if this commit gives the expected results.

I also have to investigate why tests are failing for go 1.17.

Hopefully we canmere this soon.

Reply:
I just integrated the new benchmarks you wrote, and, good news, the latest changes I made have the expected results.

Before the latest commit:

```
BenchmarkOpenTarThenReadFile_ManySmallFiles-12     	      13	  79690862 ns/op	 8479449 B/op	  163488 allocs/op
BenchmarkOpenTarThenReadFile_FewLargeFiles-12      	     134	   8717038 ns/op	45078965 B/op	     185 allocs/op
BenchmarkReadFile_ManySmallFiles-12                	  162428	      7505 ns/op	    9350 B/op	      18 allocs/op
BenchmarkReadFile_FewLargeFiles-12                 	     138	   8608644 ns/op	45072017 B/op	      21 allocs/op
BenchmarkOpenAndCloseFile_ManySmallFiles-12        	  144234	      7548 ns/op	    9441 B/op	      20 allocs/op
BenchmarkOpenAndCloseFile_FewLargeFiles-12         	     136	   8609137 ns/op	45071950 B/op	      23 allocs/op
BenchmarkOpenReadAndCloseFile_ManySmallFiles-12    	  129886	     10027 ns/op	   17666 B/op	      21 allocs/op
BenchmarkOpenReadAndCloseFile_FewLargeFiles-12     	      64	  17584599 ns/op	90138885 B/op	      25 allocs/op
```

Then after:

```
BenchmarkOpenTarThenReadFile_ManySmallFiles-12     	      14	  80771991 ns/op	 8487140 B/op	  163490 allocs/op
BenchmarkOpenTarThenReadFile_FewLargeFiles-12      	     134	   8748853 ns/op	45079883 B/op	     185 allocs/op
BenchmarkReadFile_ManySmallFiles-12                	  170936	      7069 ns/op	    9334 B/op	      18 allocs/op
BenchmarkReadFile_FewLargeFiles-12                 	     136	   8688447 ns/op	45071983 B/op	      21 allocs/op
BenchmarkOpenAndCloseFile_ManySmallFiles-12        	  285640	      4053 ns/op	    1164 B/op	      19 allocs/op
BenchmarkOpenAndCloseFile_FewLargeFiles-12         	  275209	      3937 ns/op	    1210 B/op	      19 allocs/op
BenchmarkOpenReadAndCloseFile_ManySmallFiles-12    	  152622	      7621 ns/op	    9386 B/op	      20 allocs/op
BenchmarkOpenReadAndCloseFile_FewLargeFiles-12     	     134	   8643943 ns/op	45070741 B/op	      22 allocs/op
```

For the "Open and Close" benchmark:
 - many small files goes ~2x faster and takes **~10x less memory**
 - few large files goes **~2000x faster** and takes **~37000x** less memory

For the "Open, Read and Close benchmark:
 - many small files goes a little faster and takes ~2x less memory
 - few large files goes ~2x faster and takes ~2x less memory

Other benchmarks stay stable, which is good.

I just need to investigate why go 1.17 doesn't agree with this, then we can merge.

Reply:
Nice! Thanks for collaborating on this! I'll take a look at the 1.17 failures, too.

Reply:
It looks like the 1.17 failures are all coming from this line: https://github.com/nlepage/go-tarfs/blob/2aa91f4482d304324c31c7a9eac03d42ddc7fc39/fs.go#L296

Will dig more tomorrow!